### PR TITLE
Key Content constraint methods

### PR DESCRIPTION
**Pull request by KernelPryanic**

### CODE REVIEW

File: gqt.go [hunk 1]
```diff
@@ -79,6 +79,191 @@ type ObjectField struct {
...
+		cc := c.(ConstraintLenLess)
+		return cc.Name(), cc.Value, nil
+	case ConstraintLenGreaterOrEqual:
+		cc := c.(ConstraintLenGreaterOrEqual)
+		return cc.Name(), cc.Value, nil
+	case ConstraintLenLessOrEqual:
+		cc := c.(ConstraintLenLessOrEqual)
+		return cc.Name(), cc.Value, nil
+	default:
+		return "", nil, fmt.Errorf("constraint of unknown type")
```
Comment (romshark) [inline gqt.go]: this `error` ain't necessary. Just return `"", nil`

File: gqt_test.go [hunk 2]
```diff
@@ -7,6 +7,363 @@ import (
...
+			input: gqt.ConstraintLenLessOrEqual{
+				Value: 404,
+			},
+			name:  "ConstraintLenLessOrEqual",
+			value: uint(404),
+		},
+	} {
+		t.Run(td.name, func(t *testing.T) {
+			name, value, err := gqt.ConstraintNameAndValue(td.input)
+			require.Nil(t, err)
```
Comment (romshark) [inline gqt_test.go]: use `require.NoError(t, err)` instead

File: gqt_test.go [hunk 3]
```diff
@@ -7,6 +7,363 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+type BasicConstraintInterface interface {
```
Comment (romshark) [inline gqt_test.go]: why not just `BasicConstraint` instead?

File: gqt_test.go [hunk 4]
```diff
@@ -7,6 +7,363 @@ import (
...
+	Name() string
+}
+
+type KeyValueConstraintInterface interface {
```
Comment (romshark) [inline gqt_test.go:10]: why not just `KeyValueConstraint` instead?

File: gqt.go [hunk 5]
```diff
@@ -79,6 +79,191 @@ type ObjectField struct {
...
+
+func (of ObjectField) Key() string {
+	return of.Name
+}
+
+func (of ObjectField) Content() Constraint {
+	return of.Value
+}
+
+func ConstraintNameAndValue(c Constraint) (string, interface{}, error) {
```
Comment (romshark) [inline gqt.go]: This function shouldn't be in the scope of this package since value fields are exported. It's a helper function for a specific use case and thus doesn't belong here. Packages are to be kept minimalistic.

File: gqt_test.go [hunk 6]
```diff
@@ -7,6 +7,363 @@ import (
...
+		t.Run(td.name, func(t *testing.T) {
+			name, value, err := gqt.ConstraintNameAndValue(td.input)
+			require.Nil(t, err)
+			require.Equal(t, td.name, name)
+			require.Equal(t, td.value, value)
+		})
+	}
+}
+
+func TestConstraintName(t *testing.T) {
```
Comment (romshark) [inline gqt_test.go]: this test is missing a case where `input: nil; expect: ""`

File: gqt_test.go [hunk 7]
```diff
@@ -7,6 +7,363 @@ import (
...
+		{
+			input:  gqt.ConstraintLenGreaterOrEqual{},
+			expect: "ConstraintLenGreaterOrEqual",
+		},
+		{
+			input:  gqt.ConstraintLenLessOrEqual{},
+			expect: "ConstraintLenLessOrEqual",
+		},
+	} {
+		t.Run(td.expect, func(t *testing.T) {
```
Comment (romshark) [inline gqt_test.go]: I'd suggest using `fmt.Sprintf("%T", td.input)` which will use the type name of `td.input` instead of `td.expect` for the test name because empty input should not be `""` but `nil`

File: gqt_test.go [hunk 8]
```diff
@@ -7,6 +7,363 @@ import (
...
+			name := td.input.Name()
+			require.Equal(t, td.expect, name)
+		})
+	}
+}
+
+func TestConstraintKeyAndValue(t *testing.T) {
```
Comment (romshark) [inline gqt_test.go:15]: This test is missing a case where `input: nil; key: ""; value: nil`. However, remove the test since the function to be tested is to be removed.

**Review by romshark — CHANGES_REQUESTED** [review]

File: gqt_test.go (hunk 3 above)
Comment (KernelPryanic) [reply gqt_test.go]: Just in case to not to interfere with a same name type

File: gqt_test.go (hunk 4 above)
Comment (KernelPryanic) [reply gqt_test.go:10]: Just in case to not to interfere with a same name type

File: gqt_test.go (hunk 6 above)
Comment (KernelPryanic) [reply gqt_test.go]: Why do we need this test case?

File: gqt_test.go (hunk 7 above)
Comment (KernelPryanic) [reply gqt_test.go]: Hm, but if we can print type using `fmt.Sprintf("%T", v)` then probably what I did is useless)

File: gqt_test.go (hunk 8 above)
Comment (KernelPryanic) [reply gqt_test.go:15]: Didn't understand why we need to test `input: nil`

File: gqt_test.go (hunk 3 above)
Comment (romshark) [reply gqt_test.go]: There's no other type with that name that I know of. Constraint types are always prefixed with `Constraint...`

File: gqt_test.go (hunk 6 above)
Comment (romshark) [reply gqt_test.go]: because you want to cover this case as well 😄

File: gqt_test.go (hunk 3 above)
Comment (KernelPryanic) [reply gqt_test.go]: I know, but possibly can be in the future) I've started with the name `ConstraintInterface` at first and then renamed to `BasicConstraintInterface` to even less interference.

File: gqt_test.go (hunk 6 above)
Comment (KernelPryanic) [reply gqt_test.go]: I mean it won't even work, as I use generics in this unit test, so `nil` doesn't have `Name()` method for sure)

File: gqt_test.go (hunk 7 above)
Comment (romshark) [reply gqt_test.go]: You can use `fmt.Sprintf` but it'll certainly be slower because it relies on runtime reflection.

File: gqt_test.go (hunk 8 above)
Comment (romshark) [reply gqt_test.go:15]: Because we want to test all possible inputs

File: gqt_test.go (hunk 3 above)
Comment (romshark) [reply gqt_test.go]: it's fine, I don't think there'll ever be a collision with this name. Keep things out of the type name that can be inferred from the actual type.

File: gqt_test.go (hunk 6 above)
Comment (romshark) [reply gqt_test.go]: Ah, you're calling an interface method. Right, then it doesn't make sense.
